keps/sig-node/4639-oci-volume-source/README.md

```diff
@@ -0,0 +1,883 @@
+   - **Use Case:** Allows the distribution of non-container content using the same infrastructure and tools developed for OCI images.
+
+**OCI Object:**
+   - Umbrella term encompassing both OCI images and OCI artifacts. It represents any object that conforms to the OCI specifications for storage and distribution.
+
+### Risks and Mitigations
+
+- **Security Risks:** Allowing direct mounting of OCI images introduces potential attack vectors. Mitigation includes thorough security reviews and
+  limiting access to trusted registries. Limiting to OCI artifacts (non-runnable content) or read-only mode may lessen the security risk.
+- **Compatibility Risks:** Ensure compatibility with existing features.
+- **Performance Risks:** Large images or artifacts could impact performance. Mitigation includes optimizations in the implementation and providing
+  guidance on best practices for users.
+
+## Design Details
+
+The new `VolumeSource` will be defined in the Kubernetes API, and the implementation will involve updating components (CRI, Kubelet, Scheduler)
+to support this source type. Key design aspects include:
+
+- API changes to introduce the new `VolumeSource` type.
+- Modifications to the Kubelet to handle mounting OCI images and artifacts.
+- Enhancements to the scheduler to recognize and properly allocate resources for this new volume type.
+- Handling image pull secrets and registry authentication.
+- Reuse existing logic from ConfigMaps for:
+  - Determining the file location on the host filesystem.
+  - Managing garbage collection of unused images and artifacts.
+  - Handling updates to the OCI image or artifact, similar to how ConfigMaps update mounted files.
+
+The following code snippet illustrates the proposed API change:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: example-pod
+spec:
+  volumes:
+  - name: oci-volume
+    oci:
+      image: "example.com/my-image:latest"
+  containers:
+  - name: my-container
+    image: busybox
+    volumeMounts:
+    - mountPath: /data
+      name: oci-volume
+```
+
+### Kubelet and CRI support for OCI artifacts
+
+Kubelet and the Container Runtime Interface (CRI) currently handle OCI images. To support OCI artifacts,
+potential enhancements may be required:
+
+**Extended Media Type Handling:**
+   - Update Kubelet and CRI to recognize and handle new media types associated with OCI artifacts.
+   - Ensure that pulling and storing these artifacts is as efficient and secure as with OCI images.
+
+**Lifecycling and Garbage Collection:**
+   - Implement custom logic for managing the lifecycle of various OCI artifacts.
+
+**Artifact-Specific Configuration:**
+   - Introduce new configuration options or CRD to handle the unique requirements of different types of OCI artifacts.
+
+**Validation:**
+   - Extend validation and security checks to cover new artifact types.
+
+**Storage Optimization:**
+   - Develop optimized storage solutions tailored for different artifact types,
```

Comment:
Not sure what we mean here. Can you provide more details? I would image the runtime would be responsible for this.

Reply:
@sallyom :point_up:

Reply:
Is this where I get to share my vision of decentralized ([IPFS](https://ipfs.tech/)) storage where users may opt-in to donate unused storage & bandwidth to address the unsustainable resource requirements of AI? 
LLMs can be chunked and distributed across the IPFS network (I've never done this, have only dreamed it so far). Has anyone proposed an IPFS CSI Plugin? 
An in-tree OCI VolumeSource with an IPFS VolumeMount provided by a CSI plugin could allow OCI volumes to be populated dynamically by a distributed IPFS system.

Reply:
interesting... suppose the container runtimes could also wire in decentralized fs for cached oci object storage.. that seems like per platform/cloud level optimization details.

Reply:
IPFS and decentralized filesystems have very bad performance, also the implication at the network level will be important, unless you have separate networks for the storage and the network applications ... and then you have a more complicated setup + the security concerns of the decentralized storage fs

Reply:
Yes, the details of how/where the artifacts are pulled should be transparent. Kubelet only specifies the pull spec and then it is up to the runtime impl to configure how it is made available.

Reply:
> up to the runtime impl to configure how it is made available

Which could then include the runtime's choice of lazy-loading implementation, pull mechanism, on-disk representation, etc.